src/main/java/redis/clients/jedis/ClientSideCache.java

```diff
@@ -1,71 +1,108 @@
 package redis.clients.jedis;
 
 import java.nio.ByteBuffer;
-import java.util.HashMap;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
-
-import redis.clients.jedis.exceptions.JedisException;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
 import redis.clients.jedis.util.SafeEncoder;
 
-public class ClientSideCache {
+public abstract class ClientSideCache {
 
-  private final Map<ByteBuffer, Object> cache;
+  private final Map<ByteBuffer, Set<Long>> keyHashes;
+  private final ReentrantReadWriteLock rwl = new ReentrantReadWriteLock();
+  private final Lock readLock = rwl.readLock();
+  private final Lock writeLock = rwl.writeLock();
 
-  public ClientSideCache() {
-    this.cache = new HashMap<>();
+  protected ClientSideCache() {
+    this.keyHashes = new ConcurrentHashMap<>();
   }
 
-  /**
-   * For testing purpose only.
-   * @param map 
-   */
-  ClientSideCache(Map<ByteBuffer, Object> map) {
-    this.cache = map;
+  protected ClientSideCache(Map<ByteBuffer, Set<Long>> keyHashes) {
+    this.keyHashes = keyHashes;
   }
 
-  public final void clear() {
-    cache.clear();
-  }
+  public abstract void invalidateAll();
+
+  protected abstract void invalidateAll(Iterable<Long> hashes);
+
+  protected abstract void put(long hash, Object value);
 
-  public final void invalidateKeys(List list) {
+  protected abstract Object get(long hash);
+
+  final void invalidateKeys(List list) {
     if (list == null) {
-      clear();
+      invalidateAll();
       return;
     }
 
-    list.forEach(this::invalidateKey);
+    Set<Long> hashes = new HashSet<>();
+    list.forEach(key -> hashes.addAll(getHashes(key)));
+    invalidateAll(hashes);
   }
 
-  private void invalidateKey(Object key) {
-    if (key instanceof byte[]) {
-      cache.remove(convertKey((byte[]) key));
-    } else {
-      throw new JedisException("" + key.getClass().getSimpleName() + " is not supported. Value: " + String.valueOf(key));
+  private Set<Long> getHashes(Object key) {
+    if (!(key instanceof byte[])) {
+      throw new AssertionError("" + key.getClass().getSimpleName() + " is not supported. Value: " + String.valueOf(key));
     }
-  }
 
-  protected void setKey(Object key, Object value) {
-    cache.put(getMapKey(key), value);
+    final ByteBuffer mapKey = makeKey((byte[]) key);
+    readLock.lock();
+    try {
+      Set<Long> hashes = keyHashes.get(mapKey);
+      return hashes != null ? hashes : Collections.emptySet();
+    } finally {
+      readLock.unlock();
+    }
   }
 
-  protected <T> T getValue(Object key) {
-    return (T) getMapValue(key);
+  final <T> T getValue(Function<CommandObject<T>, T> loader, CommandObject<T> command, String... keys) {
+
+    final long hash = getHash(command);
```

Comment:
Just to confirm that I understood right, we are doing this trick with command hashes in order to cover situations when different commands get differently formatted responses for the same Redis key, right? So we want to store all the different types of responses, hence we can't use a global cache based only on the Redis key.

Reply:
It's literally taken from the job description. Slacked you the link.